### ISSUE TITLE
reimplement YOLOv8-obb result on DOTA

### ISSUE THREAD

**Issue by tianlianghai** [ultralytics/ultralytics#13532]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, I'm trying to reimplement the official result on DOTA provided on the YOLOv8 doc, it seems that YOLOv8m-obb can reach an mAP over 0.8, since I can't find a merge tool in the repo, I write it by myself by referencing mmrotate by its format_result fuction. But I can only get an mAP of 0.78.
Can you release the code about how you get those merged result for DOTA multiscale split?

**Comment by GISScience** [comment]
Thanks for asking... 
can you please shares your trained model  with  mAP of 0.78.?

only which dota versions you trained ?? doata 1 , dota 2 or data 8??

**Comment by glenn-jocher** [comment]
@GISScience hello,

Thank you for reaching out! It's great to see your interest in reimplementing the YOLOv8-obb results on the DOTA dataset. To address your questions:

1. **Sharing Your Trained Model**: While sharing your trained model with an mAP of 0.78 could be helpful, it would be even more beneficial if you could provide a minimum reproducible example of your training and evaluation process. This will allow us to better understand the steps you took and identify any potential areas for improvement. You can find guidelines for creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

2. **DOTA Versions**: The DOTA dataset has multiple versions, including DOTA1.0, DOTA1.5, and DOTA2.0. Additionally, we have a smaller subset called DOTA8 for quick testing and debugging. Could you please specify which version of the DOTA dataset you used for your training? This information is crucial for ensuring consistency in results.

3. **Merge Tool for DOTA Multiscale Split**: It seems like you have implemented your own merge tool by referencing mmrotate. While this is a good approach, there might be nuances in the merging process that could affect the final mAP. We currently do not have an official merge tool in the Ultralytics repo, but we are working on providing more comprehensive tools and documentation. In the meantime, you might want to double-check the merging logic and ensure it aligns with the standard practices used in the community.

Here’s a quick example of how you can train a YOLOv8-obb model on the DOTA8 dataset:

```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n-obb.pt")  # load a pretrained model (recommended for training)

# Train the model
results = model.train(data="dota8.yaml", epochs=100, imgsz=640)
```

For more detailed information on the DOTA8 dataset and its usage, you can refer to the [DOTA8 documentation](https://docs.ultralytics.com/datasets/obb/dota8/).

If you have any further questions or need additional assistance, feel free to ask. We're here to help!

**Comment by tianlianghai** [comment]
> Thanks for asking... can you please shares your trained model with mAP of 0.78.?
> 
> only which dota versions you trained ?? doata 1 , dota 2 or data 8??

No, I don't mean that I trained a model with mAP of 0.78, I mean I write a test script by myself to test the official weight which should hava mAP of 0.8, since there is no test script provided in the yolov8 doc.

FYI, my test script is as following:
```python
import os
from pathlib import Path
from argparse import ArgumentParser
from ultralytics.models import YOLO
from argparse import ArgumentParser

import numpy as np
import sys
import os
from pathlib import Path
import ultralytics
import pickle
from mmrotate.datasets import DOTADataset
try:
    import tlhengine
except:
    # run pip install
    # install a remote url https://gitee.com/tianlianghai/tlh_engine.git
    os.system('pip install git+https://gitee.com/tianlianghai/tlh_engine.git')
from tlhengine import utils as tutils
from tlhengine.utils import IS_DEBUGGING



def subm(predict_result_path, save_dir,  test_imgs_path, nms_threshold=0.1, zip_dir='task1'):

    if isinstance(predict_result_path, Path):
        with open(predict_result_path, 'rb') as f:
            print('loading file')
            mmlab_results = pickle.load(f)
    else:
        mmlab_results = predict_result_path
        print("the result is already loaded")
   

        
    print(len(os.listdir(test_imgs_path)))
    dataset = DOTADataset(
        test_imgs_path,
        [
                dict(type='LoadImageFromFile'),
                dict(
                    type='MultiScaleFlipAug',
                    img_scale=(1024, 1024),
                    flip=False,
                    transforms=[
                        dict(type='RResize'),
                        dict(
                            type='Normalize',
                            mean=[123.675, 116.28, 103.53],
                            std=[58.395, 57.12, 57.375],
                            to_rgb=True),
                        dict(type='Pad', size_divisor=32),
                        dict(type='DefaultFormatBundle'),
                        dict(type='Collect', keys=['img'])
                    ])
            ], 
        'le90',
        img_prefix = test_imgs_path,
        test_mode=True
    )
    print(len(dataset))
    dataset.format_results(mmlab_results, save_dir, nms_threshold=nms_threshold)


def yolo_class2mmlab_class(yolo_cls_index):
    yolo_names = {0: 'plane',
    1: 'ship',
    2: 'storage tank',
    3: 'baseball diamond',
    4: 'tennis court',
    5: 'basketball court',
    6: 'ground track field',
    7: 'harbor',
    8: 'bridge',
    9: 'large vehicle',
    10: 'small vehicle',
    11: 'helicopter',
    12: 'roundabout',
    13: 'soccer ball field',
    14: 'swimming pool'}

    mmlab_names = ('plane', 'baseball-diamond', 'bridge', 'ground-track-field',
            'small-vehicle', 'large-vehicle', 'ship', 'tennis-court',
            'basketball-court', 'storage-tank', 'soccer-ball-field',
            'roundabout', 'harbor', 'swimming-pool', 'helicopter')


    # map yolo class index to mmlab class index

    # first, replace the space in yolo names 
    yolo_names = {k: v.replace(' ', '-') for k, v in yolo_names.items()}
    return mmlab_names.index(yolo_names[yolo_cls_index])

def yolo_result2mmlab_pickle(yolo_results):
    

    mmlab_results = []  # should be a list of length(images, 15 classes)
    # the item in it is array of shape (num_dets, 6), where 6 is x_center, y_center, w, h, angle, score

    for result in yolo_results:
        # a result corresponds to a image, and contains, class, confidence and xywhr, it only
        # shows dets,
        # create an empty list to store 15 empty list.
        this_img_result = []
        for i in range(15):
            this_img_result.append([])
        for i in range(len(result.obb.cls)):
            cls = yolo_class2mmlab_class((int(result.obb.cls[i])))
            conf = result.obb.conf[i].item()
            x_center, y_center, w, h, angle = result.obb.xywhr[i].cpu().tolist()
            # convert [0, pi) to [-pi/2, pi/2)
            if angle > np.pi/2:
                angle -= np.pi
                print("warning: angle > pi/2")
            this_img_result[cls].append([x_center, y_center, w, h, angle, conf])

        # covert all the list to numpy array
        for i in range(15):
            if len(this_img_result[i]) ==0 :
                this_img_result[i] = np.empty((0, 6))
            else:
                this_img_result[i] = np.array(this_img_result[i])
        mmlab_results.append(this_img_result)
    return mmlab_results

def predict(weight, save_result_path:Path,  test_imgs_path = '../datasets/DOTAv1-ms-1024-500/images/test', device=None):
    # parser = ArgumentParser()
    # parser.add_argument('--model', type=str, default='yolov8m-obb.pt')
    # parser.add_argument('--predict-result', type=str, default='mmlab_results.pkl')
    # args = parser.parse_args()
  


    # Load a model
    model = YOLO(weight)  # load an official model
   
    # test_imgs_path = '/Data1/home/tianlianghai/exp1/datasets/DOTAv1-ms-1024-500/images/debug'
   
        
    results = model.predict(test_imgs_path, half=True, batch=1, device=device)
    mmlab_results = yolo_result2mmlab_pickle(results)

    # tutils.pickle.
    # save it to pickle
    # import pickle
    tutils.save_pickle_with_obj(mmlab_results, save_result_path.joinpath('mmlab_result.pkl'))
    tutils.save_pickle_with_obj(results, save_result_path.joinpath('yolo_result.pkl'))
    
    return mmlab_results, results

def get_predict_path(weight_path, img_src_path):
    train_dir = weight_path.parents[1]
    predict_result_path = train_dir.joinpath('predict')
    # i = 2
    # while predict_result_path.exists():
    #     predict_result_path = train_dir.joinpath(f'predict{i}')
    #     i += 1
    #
    
    # get the last item of img_src_path
    # so that don't need to predict the same image again
    img_src_path = img_src_path.split('/')
    img_src_path = img_src_path[-1]
    predict_result_path = predict_result_path.joinpath(img_src_path)
    os.makedirs(predict_result_path, exist_ok=True)
    return predict_result_path

def main():
    parser = ArgumentParser()
    parser.add_argument('weight', type=str)
    parser.add_argument('--pkl-path', type=str)
    parser.add_argument('--device', type=str)
    parser.add_argument('--nms-thresh', type=float, default=0.1)
    parser.add_argument('--zip-dir', default='task1')
    args = parser.parse_args()
  

    test_img_path = '../datasets/DOTAv1-ms-1024-500/images/' + ('debug' if IS_DEBUGGING else 'test')
   
      
    # if args.skip_predict:
    #     predict_path=Path(args.pkl_path).parents[0] 
    #     subm(Path(args.pkl_path),  predict_path.joinpath('task1'),)
        
    # else:
        
    WEIGHT_PATH = args.weight
    WEIGHT_PATH = Path(WEIGHT_PATH)
    # train_dir = WEIGHT_PATH.parents[2]
    predict_path=get_predict_path(WEIGHT_PATH, test_img_path)
    test_img_path = '../datasets/DOTAv1-ms-1024-500/images/test'
    if IS_DEBUGGING:
        test_img_path = '../datasets/DOTAv1-ms-1024-500/images/debug'
    if os.path.isfile(predict_path.joinpath('yolo_result.pkl')): 
        print(f"yolo_result.pkl already exists in {predict_path}")
        print("skip predicting")
        subm(predict_path.joinpath('mmlab_result.pkl'),  predict_path.joinpath(args.zip_dir),test_imgs_path=test_img_path
         , nms_threshold=args.nms_thresh, zip_dir = args.zip_dir)

        # exit()

    # log current weight path
    else:
        with open(predict_path.joinpath('log.txt'), 'w') as f:
            f.write(str(WEIGHT_PATH))
        mmlab_result, yolo_result = predict(WEIGHT_PATH, save_result_path=predict_path,
            test_imgs_path=test_img_path, device=args.device)
    # if IS_DEBUGGING:
    #     for result in yolo_result:
    #         print(result.obb.xyxyxyxy)
        subm(mmlab_result,  predict_path.joinpath(args.zip_dir),test_imgs_path=test_img_path
            , nms_threshold=args.nms_thresh, zip_dir = args.zip_dir)


if __name__=='__main__':
    main()
```

**Comment by tianlianghai** [comment]
<img width="730" alt="image" src="https://github.com/ultralytics/ultralytics/assets/75210582/141e2ca0-be69-40ab-aea5-aad1d62d9d00">
https://docs.ultralytics.com/tasks/obb/
the tested model is YOLOv8m-obb

**Comment by glenn-jocher** [comment]
Hello @tianlianghai,

Thank you for providing detailed information and sharing your test script! It's great to see your proactive approach in reimplementing the YOLOv8-obb results on the DOTA dataset.

To address your query regarding the mAP discrepancy, here are a few suggestions:

1. **Ensure Latest Versions**: Please verify that you are using the latest versions of `torch` and `ultralytics`. You can upgrade your packages using:
    ```bash
    pip install --upgrade torch ultralytics
    ```

2. **Minimum Reproducible Example**: Your script is quite comprehensive, but if you encounter any issues, please ensure it aligns with our [minimum reproducible example guidelines](https://docs.ultralytics.com/help/minimum_reproducible_example). This helps us reproduce and investigate any potential bugs effectively.

3. **Merging Logic**: The merging logic for DOTA multiscale split can be intricate. Double-check your implementation against standard practices. Ensure that the angle normalization and class mapping are correctly handled, as these can significantly impact the final mAP.

4. **Evaluation Metrics**: Make sure that the evaluation metrics and thresholds used in your script match those used in the official evaluation. Even slight differences can lead to variations in mAP.

5. **Sample Code**: Here’s a snippet to ensure you’re on the right track with the prediction and evaluation:
    ```python
    from ultralytics import YOLO

    # Load the model
    model = YOLO("yolov8m-obb.pt")

    # Predict on test images
    results = model.predict("path/to/test/images", half=True, batch=1)

    # Convert results to the required format and evaluate
    mmlab_results = yolo_result2mmlab_pickle(results)
    subm(mmlab_results, "path/to/save/results", "path/to/test/images", nms_threshold=0.1)
    ```

If you continue to face issues, please share any specific errors or unexpected behaviors you encounter. We're here to help!
